File: c/tests/test_bitfield_flag_selector_33A98.c
/* test_bitfield_flag_selector_33A98.c
 *
 * Host C companion for bitfield_flag_selector_33A98 (0x33A98).
 *
 * The lift reads a status byte at 0xFFFFCD4E and writes a select code into
 * the top nibble of byte@0xFFFFC05C.  Both SFRs are mapped with mmap.
 *
 * Reference: v = (b & 0x40) ? 0 : (b & 0x20) ? 1 : (b & 0x80) ? 2 : 3;
 *            out = v << 4.
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>

extern void bitfield_flag_selector_33A98(void);

#define IN_BYTE  0xFFFFCD4Eu  /* input  status byte      */
#define OUT_BYTE 0xFFFFC05Cu  /* output select code byte */

static void map_page(uintptr_t addr)
{
    long page = sysconf(_SC_PAGESIZE);
    uintptr_t base = addr & ~((uintptr_t)page - 1);
    void *p = mmap((void *)base, (size_t)page, PROT_READ | PROT_WRITE,
                   MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0);
    if (p == MAP_FAILED) {
        perror("mmap");
        exit(1);
    }
}

static uint8_t ref(uint8_t b)
{
    uint8_t v = (b & 0x40) ? 0 : (b & 0x20) ? 1 : (b & 0x80) ? 2 : 3;
    return (uint8_t)(v << 4);
}

int main(void)
{
    unsigned failures = 0, tests = 0;

    map_page(IN_BYTE);
    map_page(OUT_BYTE);

    printf("=== bitfield_flag_selector_33A98 ===\n");

    for (unsigned b = 0; b < 256; b++) {
        *(volatile uint8_t *)IN_BYTE = (uint8_t)b;
        *(volatile uint8_t *)OUT_BYTE = 0xFF;  /* poison */
        bitfield_flag_selector_33A98();
        uint8_t got = *(volatile uint8_t *)OUT_BYTE;
        uint8_t exp = ref((uint8_t)b);
        tests++;
        if (got != exp) {
            printf("FAIL: in=0x%02X out=0x%02X expected 0x%02X\n", b, got, exp);
            failures++;
        }
    }

    printf("Results: %u tests, %u failures\n", tests, failures);
    return failures ? 1 : 0;
}

File: c/tests/test_bitfield_flag_selector_33A98.py
#!/usr/bin/env python3
"""
Verify bitfield_flag_selector_33A98 (0x33A98) against the ACTUAL ROM bytes,
run in the SH-2E emulator.

Side-effect leaf: reads status byte at 0xFFFFCD4E, writes select code into
the top nibble of byte@0xFFFFC05C:

  b = byte@0xFFFFCD4E
  v = (b & 0x40) ? 0 : (b & 0x20) ? 1 : (b & 0x80) ? 2 : 3
  byte@0xFFFFC05C = v << 4

C:
  void bitfield_flag_selector_33A98(void)

NOTE: mov.w sign-extends 0xCD4E → input is 0xFFFFCD4E.

Run from repo root:  python3 c/tests/test_bitfield_flag_selector_33A98.py
"""
import os, sys

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(ROOT, 'tools'))
from sh2emu import SH2

ROM = os.path.join(ROOT, 'roms', 'stock', '60E1D400.bin')
ENTRY = 0x0033A98

IN_BYTE = 0xFFFFCD4E   # input  status byte
OUT_BYTE = 0xFFFFC05C  # output select code byte


def model(b):
    v = 0 if (b & 0x40) else (1 if (b & 0x20) else (2 if (b & 0x80) else 3))
    return (v << 4) & 0xFF


def main():
    rom = open(ROM, 'rb').read()
    cpu = SH2(rom)

    for b in range(256):
        cpu.call(ENTRY, ram={IN_BYTE: b})
        got = cpu.ram.get(OUT_BYTE, -1)
        exp = model(b)
        if got != exp:
            print("FAIL: in=0x%02X -> out=0x%02X expected 0x%02X" % (b, got, exp))
            sys.exit(1)

    print("OK  bitfield_flag_selector_33A98 @0x%04X  (exhaustive 256)" % ENTRY)
    sys.exit(0)


if __name__ == '__main__':
    main()
